src/DataStreams/RemoteQueryExecutor.cpp

```diff
@@ -176,6 +185,35 @@ Block RemoteQueryExecutor::read()
             return {};
     }
 
+    try
+    {
+        return readPackets();
+    }
+    catch (const DB::Exception & ex)
+    {
+        if (ex.code() == ErrorCodes::DUPLICATED_FINGERPRINTS)
+        {
+            /// Cancel previous query and disconnect before retry.
+            cancel();
+            multiplexed_connections->disconnect();
+
+            if (!resent_query)
+            {
+                if (log)
+                    LOG_DEBUG(log, "found duplicate fingerprints, will retry query without those parts");
+                resent_query = true;
+                sent_query = false;
+                /// Consecutive read will implicitly send query first.
+                return read();
```

Comment:
Not clear. As far as I understood, if we get duplicated fingerprints we will cancel the whole query and try to execute it again? Also, logic based on exception is not the best way to do it.

Reply:
>As far as I understood, if we get duplicated fingerprints we will cancel the whole query and try to execute it again?

no no, the idea here is to retry query only against the replica whose expected result would comprised one of already processed fingerprints. hence avoid double spending of resource and don't expose fingerprints to end user.